components/locid_transform/src/directionality.rs

```diff
@@ -0,0 +1,186 @@
+    }
+
+    // Note: This is a custom impl because the bounds on `try_new_unstable` don't suffice
+    #[doc = icu_provider::gen_any_buffer_docs!(BUFFER, icu_provider, Self::try_new_unstable)]
+    #[cfg(feature = "serde")]
+    pub fn try_new_with_buffer_provider(
+        provider: &(impl BufferProvider + ?Sized),
+    ) -> Result<LocaleDirectionality, LocaleTransformError> {
+        let expander = LocaleExpander::try_new_with_buffer_provider(provider)?;
+        Self::try_new_with_expander_unstable(&provider.as_deserializing(), expander)
+    }
+
+    /// Creates a [`LocaleDirectionality`] with a custom [`LocaleExpander`] object.
+    ///
+    /// For example, use this constructor if you wish to support all languages.
+    ///
+    /// [📚 Help choosing a constructor](icu_provider::constructors)
+    /// <div class="stab unstable">
+    /// ⚠️ The bounds on this function may change over time, including in SemVer minor releases.
+    /// </div>
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_locid::locale;
+    /// use icu_locid_transform::{Direction, LocaleDirectionality, LocaleExpander};
+    ///
+    /// let ld_default = LocaleDirectionality::try_new_unstable(&icu_testdata::unstable())
+    ///     .expect("create failed");
+    ///
+    /// assert_eq!(ld_default.get(&locale!("jbn")), None);
+    ///
+    /// let expander = LocaleExpander::try_new_extended_unstable(&icu_testdata::unstable())
+    ///     .expect("create failed");
+    /// let ld_extended = LocaleDirectionality::try_new_with_expander_unstable(
+    ///         &icu_testdata::unstable(),
+    ///         expander,
+    ///     ).expect("create failed");
+    ///
+    /// assert_eq!(ld_extended.get(&locale!("jbn")), Some(Direction::RightToLeft));
+    /// ```
+    pub fn try_new_with_expander_unstable<P>(
+        provider: &P,
+        expander: LocaleExpander,
+    ) -> Result<LocaleDirectionality, LocaleTransformError>
+    where
+        P: DataProvider<ScriptDirectionV1Marker> + ?Sized,
+    {
+        let script_direction = provider.load(Default::default())?.take_payload()?;
+
+        Ok(LocaleDirectionality {
+            script_direction,
+            expander,
+        })
+    }
+
+    /// Returns the script direction of the given locale.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_locid::locale;
+    /// use icu_locid_transform::{Direction, LocaleDirectionality};
+    ///
+    /// let ld = LocaleDirectionality::try_new_unstable(&icu_testdata::unstable())
+    ///     .expect("create failed");
+    ///
+    /// assert_eq!(ld.get(&locale!("en-US")), Some(Direction::LeftToRight));
+    ///
+    /// assert_eq!(ld.get(&locale!("ar")), Some(Direction::RightToLeft));
+    ///
+    /// assert_eq!(ld.get(&locale!("foo")), None);
+    /// ```
+    pub fn get(&self, locale: &Locale) -> Option<Direction> {
+        let script = locale.id.script.or_else(|| {
```

Comment:
Suggestion: you can either simply call `expander.maximize()`, or if you think you need this special code path for performance reasons, refactor this into a (potentially public) function on LocaleExpander called `get_likely_script()` or similar.

Reply:
+1

also we should clearly document that it is applying expansion, and perhaps offer a separate method that *only* takes a script code

Reply:
> perhaps offer a separate method that only takes a script code

I think this might be confusing because people will try to call this with `locale.id.script` and then wonder how to resolve the `Option`. It's easy to convert a `Script` into a `Locale` to call this.

Expansion should be documented in a way that's not confusing to people who think direction is a property of the language, which I assume is the majority of our clients (the original reporter assumed this).

Reply:
> It's easy to convert a Script into a Locale to call this.

Is `Some(my_script).into()` the best way? I'll document that for users that have a script and want to get its direction.

Reply:
Yes

Reply:
Changed documentation to make this explicit, thoughts?